File: src/main/java/com/nevakanezah/horseenhancer/data/SecretHorses.java
package com.nevakanezah.horseenhancer.data;

import org.bukkit.EntityEffect;
import org.bukkit.Location;
import org.bukkit.attribute.Attribute;
import org.bukkit.entity.AbstractHorse;
import org.bukkit.entity.EntityType;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import net.md_5.bungee.api.ChatColor;

public class SecretHorses {
	
	private SecretHorses() {
		super();
	}
	
	/**
	 * Spawn a horse whose parents were maybe a little too close.
	 * @param loc Location to spawn the new entity
	 * @param horseData Data container to be attached to the new entity
	 * @return the newly-spawned horse
	 */
	public static AbstractHorse spawnInbred(Location loc, HorseData horseData) {
		AbstractHorse horse = (AbstractHorse)loc.getWorld().spawnEntity(loc, EntityType.ZOMBIE_HORSE);
		horseData.setUniqueID(horse.getUniqueId());
		horseData.setType(EntityType.ZOMBIE_HORSE);
		
		horse.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED).setBaseValue(0.1);
		horse.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(10);
		horse.getAttribute(Attribute.HORSE_JUMP_STRENGTH).setBaseValue(0.3);
		horse.setTamed(true);
		horse.setAge(-9999999);
		horse.setAgeLock(true);
		horse.setBreed(false);
		
		horseData.setGender("INBRED");
		
		return horse;
	}
	
	public static AbstractHorse spawnMaximule(Location loc, HorseData horseData) {
		AbstractHorse horse = (AbstractHorse)loc.getWorld().spawnEntity(loc, EntityType.MULE);
		horseData.setUniqueID(horse.getUniqueId());
		horseData.setType(EntityType.MULE);
		
		horse.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED).setBaseValue(0.35);
		horse.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(35);
		horse.getAttribute(Attribute.HORSE_JUMP_STRENGTH).setBaseValue(1.18);
		horse.setCustomName(ChatColor.DARK_RED + "MA"+ ChatColor.GOLD + "XI" + ChatColor.DARK_BLUE + "MU" + ChatColor.DARK_GREEN + "LE");
		horse.setBreed(false);
		horse.setAge(0);
		horse.setAgeLock(true);
		horse.playEffect(EntityEffect.FIREWORK_EXPLODE);
		
		horseData.setGender("UNIQUE");
		
		return horse;
	}
	
	public static AbstractHorse spawnInvincible(Location loc, HorseData horseData) {
		AbstractHorse horse = (AbstractHorse)loc.getWorld().spawnEntity(loc, EntityType.HORSE);
		horseData.setUniqueID(horse.getUniqueId());
		horseData.setType(EntityType.HORSE);
		
		horse.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED).setBaseValue(0.38);
		horse.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(15);
		horse.getAttribute(Attribute.HORSE_JUMP_STRENGTH).setBaseValue(0.565);
		horse.setCustomName(ChatColor.DARK_BLUE + "Invincible");
		horse.setAge(0);
		horse.setAgeLock(true);
		horse.setTamed(true);
		horse.setBreed(false);
		horse.playEffect(EntityEffect.FIREWORK_EXPLODE);
		PotionEffect invis = new PotionEffect(PotionEffectType.INVISIBILITY, 2147483000, 1, false, false);
		horse.addPotionEffect(invis, true);
		horseData.setGender("UNIQUE");
		
		return horse;
	}

}

File: src/main/java/com/nevakanezah/horseenhancer/util/SecretHorses.kt
package com.nevakanezah.horseenhancer.util

import com.nevakanezah.horseenhancer.database.table.Horse
import com.nevakanezah.horseenhancer.model.HorseGender
import com.nevakanezah.horseenhancer.util.HorseUtil.jumpStrengthAttribute
import com.nevakanezah.horseenhancer.util.HorseUtil.maxHealthAttribute
import com.nevakanezah.horseenhancer.util.HorseUtil.speed
import net.md_5.bungee.api.ChatColor
import org.bukkit.*
import org.bukkit.entity.AbstractHorse
import org.bukkit.entity.EntityType
import org.bukkit.entity.Firework
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

object SecretHorses {
    internal val nameSecretMule = ChatColor.DARK_RED.toString() + "MA" + ChatColor.GOLD + "XI" + ChatColor.DARK_BLUE + "MU" + ChatColor.DARK_GREEN + "LE"
    internal val nameSecretHorse = ChatColor.DARK_BLUE.toString() + "Invincible"

    @JvmOverloads
    fun spawnInbred(location: Location, horseData: Horse = Horse {}): Pair<Horse, AbstractHorse> {
        val world = location.world ?: Bukkit.getWorlds()[0]

        val horseEntity = world.spawnEntity(location, EntityType.ZOMBIE_HORSE) as AbstractHorse
        horseEntity.apply {
            speed = 0.1
            maxHealthAttribute = 10.0
            jumpStrengthAttribute = 0.3
            @Suppress("DEPRECATION")
            setBreed(false)
            age = Int.MIN_VALUE
            ageLock = true
            isTamed = true
        }

        horseData.apply {
            uid = horseEntity.uniqueId.toString()
            gender = HorseGender.INBRED
        }

        return horseData to horseEntity
    }

    @JvmOverloads
    fun spawnMaximule(location: Location, horseData: Horse = Horse {}): Pair<Horse, AbstractHorse> {
        val world = location.world ?: Bukkit.getWorlds()[0]

        val horseEntity = world.spawnEntity(location, EntityType.MULE) as AbstractHorse
        horseEntity.apply {
            speed = 0.35
            maxHealthAttribute = 35.0
            jumpStrengthAttribute = 1.18
            customName = nameSecretMule
            @Suppress("DEPRECATION")
            setBreed(false)
            age = 0
            ageLock = true
            spawnFireworks(location, world)
        }

        horseData.apply {
            uid = horseEntity.uniqueId.toString()
            gender = HorseGender.UNIQUE
        }

        return horseData to horseEntity
    }

    @JvmOverloads
    fun spawnInvincible(location: Location, horseData: Horse = Horse {}): Pair<Horse, AbstractHorse> {
        val world = location.world ?: Bukkit.getWorlds()[0]

        val horseEntity = world.spawnEntity(location, EntityType.HORSE) as AbstractHorse
        horseEntity.apply {
            speed = 0.38
            maxHealthAttribute = 15.0
            jumpStrengthAttribute = 0.565
            customName = nameSecretHorse
            @Suppress("DEPRECATION")
            setBreed(false)
            age = 0
            ageLock = true
            isTamed = true
            spawnFireworks(location, world)
            addPotionEffect(PotionEffect(PotionEffectType.INVISIBILITY, Int.MAX_VALUE, 1, false, false))
        }

        horseData.apply {
            uid = horseEntity.uniqueId.toString()
            gender = HorseGender.UNIQUE
        }

        return horseData to horseEntity
    }

    private fun spawnFireworks(location: Location, world: World) {
        (world.spawnEntity(location, EntityType.FIREWORK) as Firework).apply {
            fireworkMeta = fireworkMeta.apply {
                addEffect(FireworkEffect.builder()
                    .with(FireworkEffect.Type.BALL)
                    .withColor(
                        ChatColor.DARK_RED.bukkitColor,
                        ChatColor.GOLD.bukkitColor,
                        ChatColor.DARK_BLUE.bukkitColor,
                        ChatColor.DARK_GREEN.bukkitColor,
                    )
                    .withFlicker()
                    .build())
                power = 0
            }
        }
    }

    private val ChatColor.bukkitColor
        get() = this.color?.let { color -> Color.fromBGR(color.rgb and 0xff_ff_ff) } ?: Color.WHITE
}
